### DISCUSSION TITLE
How to preserve value in form when an action got validation errors?

### DISCUSSION THREAD

**Discussion by armandabric** [next-safe-action/next-safe-action#296] [Q&A]
Hey there!

I'm struggling to use `next-safe-action` with a form that could be rejected with validation errors. 

When the form is submit Next 15 (or React 19?) reset it, and I could not find a way to prevent it to became empty. After a submition that trigger validation errors, the `action.result` entry is only filled with the `validationErrors`. It seems there is no way to send back the data back like advised now with the react `useActionState`.

I have created a demo to illustrate my problem: https://codesandbox.io/p/devbox/rv6f4c

Is someone have an idea on how to do?

**Comment by SwiichyCode** [comment]
Did you find a solution ?

**Reply by armandabric** [reply]
As workaround, I remove the `action` and use `onSubmit` on the form: 


```ts
      onSubmit={(event) => {
        event.preventDefault();

        myAction.execute(new FormData(event.currentTarget));
      }}
```

Its not a "good solution" as it bypass all the React 19 stuff but it works

**Reply by alimek** [reply]
hmm I am trying to get solution for this problem too, any better solution for this guys?

I still wanna schema validation happening 
```typescript

const schema = zdf.formType({
  name: zdf.text()
});

clientAction
  .schema(schema)
  .safeAction<{ name: string }>(async ({ parsedInput }) => {
    ...

    return {
         name: ...
    };
  })
```
but I would like to also return the client input data as part of same response, so i can pass it to `defaultValue={result?.name}` to my text `input` component.

I am out of ideas how I can be achieved (if even possible) without workaround.

@TheEdoRan sorry for pinging directly, but any idea if that is currently possible? I think I scanned doc and I cant find any info 😬

**Reply by azizk17** [reply]
You can easily include the original client input data in the `handleValidationErrorsShape` 
```ts
.schema(schema, {
    handleValidationErrorsShape: async (validationErrors, utils) => {
        return {
            ...flattenValidationErrors(validationErrors).fieldErrors,
            data: Object.fromEntries(utils.clientInput.entries()),
        };
    },
})
```

That said, there are a couple of things to keep in mind. Right now, The main issue is that the `data` is nested inside `validationErrors` instead of being at the root level.

```json
{
  "validationErrors": {
    "name": [
      "String must contain at least 5 character(s)"
    ],
    "data": {
      "name": ""
    }
  }
}
```
But ideally, it should look like this:

```json
{
  "validationErrors": {
    "name": [
      "String must contain at least 5 character(s)"
    ]
  },
  "data": {
    "name": ""
  }
}
```

**Comment by azizk17** [comment]
I found out that the input (which is formData) actually preserves the values passed into it! 

```ts
const { execute, result, status, hasErrored, isPending, input } = useStateAction(serverAction, {});
 ```

**Reply by alimek** [reply]
The problem with `input` is that if you use `zod-form-data`

```ts
const schema = zfd.formData({
  name: zfd.text(
    z.string().min(2, { message: "Name must be at least 2 characters." }),
  ),
  description: zfd.text(z.string().optional()).optional(),
});
```

 it returns `FormData | FormDataLikeInput` and you can't do 


```ts
const { execute, isPending, input, result } = useStateAction(createProject);
input.get('name');
```
as I am getting error: 

```ts
Property 'get' does not exist on type 'FormData | FormDataLikeInput'.
  Property 'get' does not exist on type 'FormDataLikeInput'.ts(2339)
```

have you maybe faced it too and handled this in some way?

**Comment by milad-mamandi** [comment]
I have the same issue as well, property get does not exist on type FormDataLikeInput
any solutions?

or is there a way to use form action without zod-form-data?

**Comment by CauaYves** [comment]
I created a function to capture the input value based on its 'name' and thus keep the values:

```
	function getField(name: string): string {
		if (!input) return "";
		if (typeof (input as FormData).get === "function") {
			const val = (input as FormData).get(name);
			return typeof val === "string" ? val : "";
		}
		return (input as Record<string, string>)[name] || "";
	}
```

`<Input
	name="name"
	label="Nome completo"
	autoComplete="username"
	defaultValue={getField("name")}
/>`
